GuCDK Housekeeper

deedeeh opened the pull request:
## What does this change?

This PR migrates Ophan Housekeeper to GuCDK which is tracked in [#5086](https://github.com/guardian/ophan/issues/5086). The previous migrations, [Time Machine ](https://github.com/guardian/ophan/pull/5126), [Dashboard](https://github.com/guardian/ophan/pull/5146) and [Poller](https://github.com/guardian/ophan/pull/5174), we used `GuEc2App` but Housekeeper is a lambda and that is why we used `GuLambdaFunction`. 

We paired with @akash1810 in the DevX team as this was our first use of Lambda in GuCDK.

## How to test

We will update AWS CloudFormation with the GuCDK generated CFN and merge to main then we are going to compare the [logs in ELK](https://logs.gutools.co.uk/s/ophan/app/discover#/?_a=(columns:!(message,user.id),filters:!(),index:'581f7b00-c4df-11e9-bc08-9d1af4d9c1d5',interval:auto,query:(language:kuery,query:'%22running%20with%20notification%20message%22'),sort:!(!('@timestamp',desc)))&_g=(filters:!(),refreshInterval:(pause:!f,value:10000),time:(from:now-12h,to:now))) with the [CloudWatch metrics for the SES](https://eu-west-1.console.aws.amazon.com/cloudwatch/home?region=eu-west-1#metricsV2:graph=~(metrics~(~(~'AWS*2fSES~'Bounce)~(~'.~'Delivery~(visible~false))~(~'.~'Reputation.BounceRate~(visible~false))~(~'.~'Reputation.ComplaintRate~(stat~'Average~visible~false))~(~'.~'Send~(stat~'Average~visible~false)))~view~'timeSeries~stacked~false~region~'eu-west-1~stat~'Sum~period~300~start~'-PT12H~end~'P0D);query=~'*7bAWS*2fSES*7d*20SES) and they are suppose to match if things are working well otherwise we will see bounces in CloudWatch that doesn't appear in our logs. We should be able to know the following day if we have a problem or things work as expected.

## How can we measure success?

We have the Ophan Housekeeper using GuCDK!

## Have we considered potential risks?

The risk is very low in here because the last time The Guardian was in an SES `probation` period, which means being added to a deny list and preventing The Guardian from sending emails, happened when the Lambda wasn't working for almost a month (for more info check this [PR](https://github.com/guardian/ophan/issues/2765)) but in this PR case we will check the following day if everything works as it should be and if not we will figure out what is missing and redo the same process again. One last thing to mention is AWS will notify us before we end up in `probation` and we have a grace period to resolve the issue - check [AWS SES docs](https://aws.amazon.com/blogs/messaging-and-targeting/what-if-you-get-an-ses-probation-notice-for-bounces/)

akash1810 commented on cdk/lib/housekeeper.ts:
```diff
@@ -0,0 +1,70 @@
+
+		const app = 'housekeeper';
+
+		const loggingPolicy = new PolicyStatement();
+		loggingPolicy.addResources('arn:aws:logs:*:*:*');
+		loggingPolicy.addActions(
+			'logs:CreateLogGroup',
+			'logs:CreateLogStream',
+			'logs:PutLogEvents',
+		);
```
minor: this could a _little_ terser via the constructor props:

```ts
const loggingPolicy = new PolicyStatement({
  actions: [
    'logs:CreateLogGroup',
		'logs:CreateLogStream',
		'logs:PutLogEvents'
  ],
  resources: [
    'arn:aws:logs:*:*:*'
  ],
});
```

deedeeh replied:
I was thinking if there was a similar way of having those props inside but I wasn't sure. That looks better! Done the change 👍🏼

akash1810 commented on cdk/lib/housekeeper.ts:
```diff
@@ -0,0 +1,70 @@
+		//TODO: change the codebase to use Dev-ophan-alerts table when testing locally or in CODE or remove this table if not used
+		const dynamodbPolicy = new PolicyStatement();
+		dynamodbPolicy.addResources(
+			`arn:aws:dynamodb:*:${this.account}:table/ophan-alerts`,
+			`arn:aws:dynamodb:*:${this.account}:table/DEV-ophan-alerts`,
```
One for a future PR, but we should scope these ARNs to a specific region (`eu-west-1`?!), rather than `*`.

deedeeh replied:
Is there a specific reason to be for a future PR? Why not in this one?

akash1810 replied:
You could do - I just wasn't sure if `*` in the YAML was intentional or not.

deedeeh replied:
Me too I am not sure but I can tag @rtyley because he would know the answer for that question. I usually see the `*` for resources items or actions but not region.

deedeeh replied:
We just replaced it with the specific region during my pairing session with @SHession

SHession reviewed (approved):
This looks great, we should be good to go! 
A note about the worst case scenario, if we do end-up in probation in SES, AWS will notify us and we have a grace period to resolve the issue:

> So you’re sending lots of email; everything is looking great.  At least, until you get a probation notice in your inbox.  How could this be?  Why did it happen?  What are your next steps?
> 
> First of all, take a deep breath.  There is generally a grace period for addressing issues you might be having with your sending.

– [AWS SES Docs](https://aws.amazon.com/blogs/messaging-and-targeting/what-if-you-get-an-ses-probation-notice-for-bounces/)

deedeeh commented:
> This looks great, we should be good to go! A note about the worst case scenario, if we do end-up in probation in SES, AWS will notify us and we have a grace period to resolve the issue:
> 
> > So you’re sending lots of email; everything is looking great.  At least, until you get a probation notice in your inbox.  How could this be?  Why did it happen?  What are your next steps?
> > First of all, take a deep breath.  There is generally a grace period for addressing issues you might be having with your sending.
> 
> – [AWS SES Docs](https://aws.amazon.com/blogs/messaging-and-targeting/what-if-you-get-an-ses-probation-notice-for-bounces/)

Thank you Sam! I have read that before yes so it is good that you shared the docs with me and reminded me of that point. I added it to the potential risk question.
